sql/core/src/test/scala/org/apache/spark/sql/CachedTableSuite.scala

```diff
@@ -655,6 +663,148 @@ class CachedTableSuite extends QueryTest with SQLTestUtils with SharedSQLContext
     }
   }
 
+  test("SPARK-19993 subquery caching") {
+    withTempView("t1", "t2") {
+      Seq(1).toDF("c1").createOrReplaceTempView("t1")
+      Seq(1).toDF("c1").createOrReplaceTempView("t2")
+
+      val ds1 =
+        sql(
+          """
+            |SELECT * FROM t1
+            |WHERE
+            |NOT EXISTS (SELECT * FROM t1)
+          """.stripMargin)
+      assert(getNumInMemoryRelations(ds1) == 0)
+
+      ds1.cache()
+
+      val cachedDs =
+        sql(
+          """
+            |SELECT * FROM t1
+            |WHERE
+            |NOT EXISTS (SELECT * FROM t1)
+          """.stripMargin)
+      assert(getNumInMemoryRelations(cachedDs) == 1)
+
+      // Additional predicate in the subquery plan should cause a cache miss
+      val cachedMissDs =
+        sql(
+          """
+            |SELECT * FROM t1
+            |WHERE
+            |NOT EXISTS (SELECT * FROM t1 where c1 = 0)
+          """.stripMargin)
+      assert(getNumInMemoryRelations(cachedMissDs) == 0)
+
+      // Simple correlated predicate in subquery
+      val ds2 =
+        sql(
+          """
+            |SELECT * FROM t1
+            |WHERE
+            |t1.c1 in (SELECT t2.c1 FROM t2 where t1.c1 = t2.c1)
+          """.stripMargin)
+      assert(getNumInMemoryRelations(ds2) == 0)
+
+      ds2.cache()
+
+      val cachedDs2 =
+        sql(
+          """
+            |SELECT * FROM t1
+            |WHERE
+            |t1.c1 in (SELECT t2.c1 FROM t2 where t1.c1 = t2.c1)
+          """.stripMargin)
+
+      assert(getNumInMemoryRelations(cachedDs2) == 1)
+
+      spark.catalog.cacheTable("t1")
+      ds1.unpersist()
```

Comment:
How about splitting the test cases to multiple individual ones? The cache will be cleaned for each test case. This can avoid any extra checking, like `assert(getNumInMemoryRelations(cachedMissDs) == 0)` or `ds1.unpersist()`
```Scala
  override def afterEach(): Unit = {
    try {
      spark.catalog.clearCache()
    } finally {
      super.afterEach()
    }
  }
```